chore: various async fixes (#527)
WebAssembly/component-model: design/mvp/Async.md

### Structured concurrency

Calling *into* a component creates a `Task` to track ABI state related to the
*callee* (like "number of outstanding borrows").

Calling *out* of a component creates a `Subtask` to track ABI state related to
the *caller* (like "which handles have been lent").

When one component calls another, there is thus a `Subtask`+`Task` pair that
collectively maintains the overall state of the call and enforces that both
components uphold their end of the ABI contract. But when the host calls into
a component, there is only a `Task` and, symmetrically, when a component calls
into the host, there is only a `Subtask`.

> Based on this, the call stack for a component to host-defined import will be of
> the general form:

Author: ```suggestion
Based on this, the call stack when a component calls a host-defined import will 
be a call stack of the general form:
```

> the general form:

Author: OK I've updated it, WDYT?

### Structured concurrency

Calling *into* a component creates a `Task` to track ABI state related to the
*callee* (like "number of outstanding borrows").

Calling *out* of a component creates a `Subtask` to track ABI state related to
the *caller* (like "which handles have been lent").

When one component calls another, there is thus a `Subtask`+`Task` pair that
collectively maintains the overall state of the call and enforces that both
components uphold their end of the ABI contract. But when the host calls into
a component, there is only a `Task` and, symmetrically, when a component calls
into the host, there is only a `Subtask`.

> Based on this, the call stack when a component calls a host-defined import will 
> be a call stack of the general form:

Reviewer: ```suggestion
Based on this, the call stack when a component calls a host-defined import will 
have the general form:
```

The stackful ABI is currently gated by the 🚟 feature.

The async stackful export function signature for export `foo` (defined above

> in world `w`) is:
> 

Reviewer: ```suggestion
in world `w`) is:
```
As a general convention, this repo doesn't have a newline between sentences ending in a colon and the codeblock right after.

Author: Ah thanks for the heads up!
